Games/Solitaire/SolitaireWidget.cpp

```diff
@@ -0,0 +1,435 @@
+            || m_animation.card()->rect().right() < 0) {
+            randomize_card();
+        }
+
+        m_animation.tick();
+    }
+
+    if (m_has_to_repaint || m_game_over_animation) {
+        m_repaint_all = false;
+        update();
+    }
+}
+
+void SolitaireWidget::randomize_card()
+{
+    srand(time(NULL));
+
+    auto card = Card::construct(static_cast<Card::Type>(rand() % Card::Type::__Count), rand() % Card::card_count);
+    card->set_position({ rand() % (SolitaireWidget::width - Card::width), rand() % (SolitaireWidget::height / 8) });
+
+    int x_sgn = card->position().x() > (SolitaireWidget::width / 2) ? -1 : 1;
+    m_animation = Animation(card, rand_float() + .4, x_sgn * ((rand() % 3) + 3), .4 + rand_float() * .6);
+}
+
+void SolitaireWidget::start_game_over_animation()
+{
+    if (m_game_over_animation)
+        return;
+
+    randomize_card();
+    m_game_over_animation = true;
+}
+
+void SolitaireWidget::stop_game_over_animation()
+{
+    if (!m_game_over_animation)
+        return;
+
+    m_game_over_animation = false;
+    m_repaint_all = true;
+    update();
+}
+
+void SolitaireWidget::setup()
+{
+    stop_game_over_animation();
+
+    for (auto& stack : m_stacks)
+        stack.clear();
+
+    NonnullRefPtrVector<Card> cards;
+    for (int i = 0; i < Card::card_count; ++i) {
+        cards.append(Card::construct(Card::Type::Clubs, i));
+        cards.append(Card::construct(Card::Type::Spades, i));
+        cards.append(Card::construct(Card::Type::Hearts, i));
+        cards.append(Card::construct(Card::Type::Diamonds, i));
+    }
+
+    srand(time(NULL));
+    for (int i = 0; i < 200; ++i)
+        cards.append(cards.take(rand() % cards.size()));
+
+    make_pile(cards, stack(Pile1), 1);
+    make_pile(cards, stack(Pile2), 2);
+    make_pile(cards, stack(Pile3), 3);
+    make_pile(cards, stack(Pile4), 4);
+    make_pile(cards, stack(Pile5), 5);
+    make_pile(cards, stack(Pile6), 6);
+    make_pile(cards, stack(Pile7), 7);
+
+    while (!cards.is_empty())
+        stack(Stock).push(cards.take_last());
+
+    m_score = 0;
+    update_score(0);
+    update();
+}
+
+void SolitaireWidget::update_score(int to_add)
+{
+    m_score = AK::max(static_cast<int>(m_score) + to_add, 0);
+    m_on_score_update(m_score);
+}
+
+void SolitaireWidget::keydown_event(GUI::KeyEvent& evt)
+{
+    if (evt.key() == KeyCode::Key_F12)
+        start_game_over_animation();
+}
+
+void SolitaireWidget::mousedown_event(GUI::MouseEvent& evt)
+{
+    GUI::Widget::mousedown_event(evt);
+
+    if (m_game_over_animation)
+        return;
+
+    auto click_location = evt.position();
+    for (auto& to_check : m_stacks) {
+        if (to_check.bounding_box().contains(click_location)) {
+            if (to_check.type() == CardStack::Type::Stock) {
+                if (stack(Waste).is_empty() && stack(Stock).is_empty())
+                    return;
+
+                auto& waste = stack(Waste);
+                auto& stock = stack(Stock);
```

Comment:
Nitpick: If you grab these before, it would look a little nicer IMO:

```cpp
auto& waste = stack(Waste);
auto& stock = stack(Stock);

if (waste.is_empty() && stock.is_empty())
    return;
```